### ISSUE TITLE
Passkey save error on Oracle Cloud

### ISSUE THREAD

**Issue by holow29** [bitwarden/clients#12060]
### Steps To Reproduce

1. Register for an account at https://www.oracle.com/cloud/sign-in.html
2. During registration, it will ask if you want to save a passkey/fido authenticator.
3. Click setup
4. Save passkey in bitwarden

If you already have an account, you can add a passkey under Profile -> Security -> Fido Passkey Authenticator -> Configure.

### Expected Result

Expected indication that passkey was successfully saved on Oracle's end

### Actual Result

Nothing/error: `TypeError: Illegal invocation`

### Additional Context

First mentioned in #6804 

### Operating System

macOS

### Web Browser

Brave

### Browser Version

1.73 (Chromium 131)

### Build Version

2024.11.0

### Issue Tracking Info

- [X] I understand that work is tracked outside of Github. A PR will be linked to this issue should one be opened to address it, but Bitwarden doesn't use fields like "assigned", "milestone", or "project" to track progress.

**Comment by gboudreau** [comment]
Of note: I'm also receiving the `TypeError: Illegal invocation` error, but when trying to log in on Oracle Cloud using my Passkey, and using 1Password. So maybe the problem is from Oracle Cloud, and now Bitwarden. TBC.

**Comment by physx2494** [comment]
I've also started receiving `Illegal Invocation` as well when using passkey already stored on Bitwarden for OCI. I'm using `Version 131.0.6778.93 (Official Build, ungoogled-chromium) (64-bit)`

**Comment by danktankk** [comment]
same error and issue.  I cant use bitwarden either

**Comment by Tvangeste** [comment]
I've digged deeper into the problem, since it started to annoy me more and more.

My take on the problem is here:  https://github.com/aws-samples/amazon-cognito-passwordless-auth/issues/94#issuecomment-2550249139

Basically, Chrome 129 added PublicKeyCredential.toJSON() function, per spec, and the credentials from Bitwarden/1Password are not ready and calling this method leads to TypeError.

I assume that the best way forward is for the Bitwarden/1Password extensions to properly implement/handle PublicKeyCredential.toJSON().

Workarounds for now: 
1. Use Chrome 128
2.  Play with devtools and local overrides to delete  PublicKeyCredential.toJSON from the object
3. Wait for Bitwarned/1Password fix their implementations according to spec.

For more info about browser compatibility with PublicKeyCredential.toJSON() take a look here:

https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredential/toJSON#browser_compatibility

Basically, every browser that implements this toJSON() function breaks authentication with 1Password (and I assume Bitwarden too).

**Comment by Qualify5** [comment]
The same issue seems to occur at https://signon.oracle.com/ using Bitwarden in Edge or Chrome. I was previously able to create a passkey and save it to Bitwarden, and it worked as expected up until a few months ago. Since then, I get the Illegal invocation error.

The steps are:
1. Enter an e-mail address at https://signon.oracle.com/ - you are redirected to https://signon-int.oracle.com/signin.
2. Click the button labelled "Verify passkey on device Bitwarden" - Bitwarden launches as expected and prompts to select a passkey.
3. Select a passkey to use - the logon page instantly displays "TypeError: Illegal invocation".

**Comment by kurumigi** [comment]
> 
> Basically, Chrome 129 added PublicKeyCredential.toJSON() function, per spec, and the credentials from Bitwarden/1Password are not ready so calling this method leads to TypeError.
> 

Oh, It solved my question.
In Firefox 133.0.3, Another error message appears, "TypeError: 'toJSON' called on an object that does not implement interface PublicKeyCredential."

**Comment by holow29** [comment]
@coroiu Is implementing PublicKeyCredential on BW roadmap?

**Comment by coroiu** [comment]
@holow29 Thanks for bringing this to my attention! I've brought this up internally and it's currently being reviewed/planned. We'll hopefully have a fix for this soon!

**Comment by holow29** [comment]
@coroiu Great to hear! Thanks for picking this up. As an aside, I've found https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredential somewhat helpful to keep up with some of the work browser devs are doing for web compat. There are some new experimental APIs and methods (e.g. signalAllAcceptedCredentials_static https://chromestatus.com/feature/5101778518147072) being tested in Chromium which might be good to keep an eye on as development progresses. Also a potentially useful test site: https://signal-api-demo.glitch.me/

**Comment by Omar007** [comment]
^ same here. I was about to make an issue when I saw this one. Same is happening here:  

> TypeError: 'toJSON' called on an object that does not implement interface PublicKeyCredential.  

It's not blocking for me as I'm using Firefox so I've just set `security.webauthn.enable_json_serialization_methods` to `false` (in `about:config`) for now. This allows use of the passkey (at least in this context) without any problems for now.

**Comment by CompeyDev** [comment]
I'm on Firefox, but for other sites this workaround does not work, and instead leads to a separate error where `PublicKeyCredential.parseCreationOptionsFromJSON` is not a function.

**Comment by noahbald** [comment]
For any other strangers who run into this, I have the following workaround
```ts
const credential: Credential = /* ... */;
const isPasskeyProvidedByAuthenticator =
  credential instanceof PublicKeyCredential &&
  credential.getClientExtensionResults !==
    PublicKeyCredential.prototype.getClientExtensionResults;
const registrationResponseJson = isPasskeyProvidedByAuthenticator
  ? JSON.stringify({ ...credential }, (_, value) => {
      if (value instanceof Uint8Array) {
        return base64Encode(value);
      }
      if (value instanceof AuthenticatorAttestationResponse) {
        // @ts-ignore this should be serialized as a property
        value.authenticatorData = value.getAuthenticatorData();
      }
      return value;
    })
  : JSON.stringify(credential);

```

**Comment by Dreamacro** [comment]
For those who want to bypass temporarily, delete toJSON in the console: `delete PublicKeyCredential.prototype.toJSON`

**Comment by Qualify5** [comment]
I use Bitwarden to store a passkey for a service that recently removed password access, so that the only way to log in is now by using a passkey. This effectively locked me out, because it wasn't possible to use the passkey using Edge or Chrome on PC in combination with the extension, because of this issue. However, I found that _was_ able to use my passkey if I used the Bitwarden Android app in combination with the Chrome Android app. Posting this in case anyone else finds themselves in the same situation.

**Comment by mzinner** [comment]
It would be great to get a fix for this as it affects not only Oracle Cloud but other sites as well. Using Dreamacro's solution of running `delete PublicKeyCredential.prototype.toJSON` in the Developer Tools console is a valid workaround on desktop machines, but painful to use on every login.

**Comment by NaNisavailable** [comment]
`delete PublicKeyCredential.prototype.toJSON` worked for me but it's annoyed that I have to do it every time

**Comment by timkgh** [comment]
How is this not a priority for Bitwarden to fix? It's been open for a while.

**Comment by florian-glombik** [comment]
I fixed it with this method, for the used helper functions you can refer to https://github.com/ls1intum/Artemis/pull/10785

```
/**
 * <p><strong>The Bitwarden Plugin does not work properly on Chrome; the returned credential is malformed and cannot be stringified.</strong></p>
 *
 * <p>Unfortunately, Bitwarden does not seem to show any interest in fixing this issue:
 * <a href="https://github.com/bitwarden/clients/issues/12060" target="_blank">https://github.com/bitwarden/clients/issues/12060</a>,
 * and relies on other developers to fix it within their applications. 😞</p>
 *
 * <p>As Bitwarden is a widely used password manager, we decided to implement a workaround in our application
 * to ensure a smooth user experience.</p>
 *
 * <p>The underlying issue is that Bitwarden's returned credential does not implement <code>toJSON()</code>,
 * which is required for <code>JSON.stringify()</code> to work. As a result, the browser cannot serialize the object properly,
 * causing credential registration and login with passkeys to fail when using the Bitwarden browser plugin in Chrome.</p>
 *
 * <p><strong>We fix the issue by creating a serializable copy of the object with its properties converted into the correct format.</strong></p>
 */
export function getCredentialFromMalformedBitwardenObject(malformedBitwardenCredential: MalformedBitwardenCredential | null): Credential | null {
    if (!malformedBitwardenCredential) {
        return null;
    }

    const clonedCredential: Credential = cloneDeep(malformedBitwardenCredential) as unknown as Credential;
    const credential = JSON.parse(JSON.stringify(clonedCredential));

    return {
        //@ts-expect-error authenticatorAttachment is a method in the (getAuthenticatorAttachment) object, but we return it as property here for simplicity, assuming that we only want to stringify it afterward anyway
        authenticatorAttachment: credential.authenticatorAttachment,
        clientExtensionResults: malformedBitwardenCredential.getClientExtensionResults(),
        id: credential.id,
        rawId: convertToBase64(credential.rawId),
        response: {
            attestationObject: convertToBase64(credential.response.attestationObject),
            authenticatorData: convertToBase64(
                malformedBitwardenCredential.response.authenticatorData ?? malformedBitwardenCredential.response.getAuthenticatorData?.() ?? undefined,
            ),
            clientDataJSON: convertToBase64(credential.response.clientDataJSON),
            publicKey: convertToBase64(malformedBitwardenCredential.response.getPublicKey?.() ?? undefined),
            publicKeyAlgorithm: malformedBitwardenCredential.response.getPublicKeyAlgorithm?.() ?? undefined,
            transports: malformedBitwardenCredential.response.getTransports?.() ?? undefined,
            signature: convertToBase64(malformedBitwardenCredential.response.signature),
            userHandle: convertToBase64(malformedBitwardenCredential.response.userHandle),
        },
        type: credential.type,
    };
}

```

**Comment by casperfrx** [comment]
While thanking you for this effort, it is still really frustrating that Bitwarden has not shown any interest in this issue. I recently introduced Bitwarden to our organisation but seeing how this doesn't at the very least get a proper reply makes me question whether to propagate business with them.

Bitwarden, it should be clear to you that when you .arket passkeys especially having launched passwordless.dev, a gamble whether your passkey implementation with will work with major companies like Oracle Cloud (which it doesn't) is not something organisations and companies can build on. If you're not serious about this, then we should not take Bitwarden as a serious option.

**Comment by holow29** [comment]
Finally some progress...guess it only took a fresh issue to be open. Really disheartening when it feels like these issues aren't even looked at. https://github.com/bitwarden/clients/pull/14734

**Comment by casperfrx** [comment]
It really is. Thanks for the heads-up and thank god.

**Comment by abergs** [comment]
Hey all! Apologies that it took a while for us to get to this. This is being adressed in https://github.com/bitwarden/clients/pull/14734.

**Comment by holow29** [comment]
@abergs I know this feature was reverted from main to do testing in a feature branch. Is there a current feature branch with this implemented? I'm having trouble finding one.
Edit: Seems like it is now available for testing: https://github.com/bitwarden/clients/pull/15028

**Comment by holow29** [comment]
Resolved in v2025.7.0
